executor/window.go

```diff
@@ -0,0 +1,160 @@
+		defer span1.Finish()
+	}
+	if e.runtimeStats != nil {
+		start := time.Now()
+		defer func() { e.runtimeStats.Record(time.Now().Sub(start), chk.NumRows()) }()
+	}
+	chk.Reset()
+	if e.meetNewGroup && e.windowFunc.HasRemainingResults() {
+		err := e.appendResult2Chunk(chk)
+		if err != nil {
+			return err
+		}
+	}
+	for !e.executed && (chk.NumRows() == 0 || chk.RemainedRows(chk.NumCols()-1) > 0) {
+		err := e.consumeOneGroup(ctx, chk)
+		if err != nil {
+			e.executed = true
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeOneGroup(ctx context.Context, chk *chunk.Chunk) error {
+	var err error
+	if err = e.fetchChildIfNecessary(ctx, chk); err != nil {
+		return errors.Trace(err)
+	}
+	for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
+		e.meetNewGroup, err = e.groupChecker.meetNewGroup(e.inputRow)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.meetNewGroup {
+			err := e.consumeGroupRows(chk)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = e.appendResult2Chunk(chk)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		e.groupRows = append(e.groupRows, e.inputRow)
+		if e.meetNewGroup {
+			e.inputRow = e.inputIter.Next()
+			return nil
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeGroupRows(chk *chunk.Chunk) error {
+	if len(e.groupRows) == 0 {
+		return nil
+	}
+	e.copyChk(chk)
+	var err error
+	e.groupRows, err = e.windowFunc.ProcessOneChunk(e.ctx, e.groupRows, chk, e.partialResult)
+	return err
+}
+
+func (e *WindowExec) fetchChildIfNecessary(ctx context.Context, chk *chunk.Chunk) (err error) {
+	if e.inputIter != nil && e.inputRow != e.inputIter.End() {
+		return nil
+	}
+
+	// Before fetching a new batch of input, we should consume the last groupChecker.
+	err = e.consumeGroupRows(chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	childResult := e.children[0].newFirstChunk()
+	err = e.children[0].Next(ctx, childResult)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.childResults = append(e.childResults, childResult)
+	// No more data.
+	if childResult.NumRows() == 0 {
+		e.executed = true
+		err = e.appendResult2Chunk(chk)
+		return errors.Trace(err)
+	}
+
+	e.inputIter = chunk.NewIterator4Chunk(childResult)
+	e.inputRow = e.inputIter.Begin()
+	return nil
+}
+
+// appendResult2Chunk appends result of the window function to the result chunk.
+func (e *WindowExec) appendResult2Chunk(chk *chunk.Chunk) error {
+	e.copyChk(chk)
+	var err error
+	e.groupRows, err = e.windowFunc.ExhaustResult(e.ctx, e.groupRows, chk, e.partialResult)
+	return err
+}
+
+func (e *WindowExec) copyChk(chk *chunk.Chunk) {
+	if len(e.childResults) == 0 || chk.NumRows() > 0 {
+		return
+	}
+	childResult := e.childResults[0]
+	e.childResults = e.childResults[1:]
+	columns := e.Schema().Columns[:len(e.Schema().Columns)-1]
+	for i, col := range columns {
+		chk.CopyColumns(childResult, i, col.Index)
```

Comment:
Can we do a shallow copy here to improve the performance? For example, call `chk.SwrapColumn()`?

Reply:
The copy columns is shallow copy. We cannot use `SwapColumns` here because of `groupRows` reference the chunk in the child, so when evaluating teh window function using `groupRows`, it may cause panic.